File: src/main/java/pers/neige/neigeitems/action/node/impl/GaussianParser.java
package pers.neige.neigeitems.action.node.impl;

import lombok.NonNull;
import lombok.val;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.utils.ListUtils;
import pers.neige.neigeitems.utils.NumberParser;
import pers.neige.neigeitems.utils.ScriptUtils;
import pers.neige.neigeitems.utils.StringUtils;

import java.math.BigDecimal;
import java.util.concurrent.ThreadLocalRandom;

public class GaussianParser extends NodeParser {
    public GaussianParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "gaussian";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        return handle(
            getParsedValue(context, params, "base"),
            getParsedValue(context, params, "spread"),
            getParsedValue(context, params, "maxSpread"),
            getParsedValue(context, params, "fixed"),
            getParsedValue(context, params, "min"),
            getParsedValue(context, params, "max"),
            getParsedValue(context, params, "mode")
        );
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull String params
    ) {
        val paramsList = StringUtils.split(params, '_', '\\', 7);
        return handle(
            paramsList.get(0),
            ListUtils.getOrNull(paramsList, 1),
            ListUtils.getOrNull(paramsList, 2),
            ListUtils.getOrNull(paramsList, 3),
            ListUtils.getOrNull(paramsList, 4),
            ListUtils.getOrNull(paramsList, 5),
            ListUtils.getOrNull(paramsList, 6)
        );
    }

    private @Nullable String handle(
        @Nullable String base,
        @Nullable String spread,
        @Nullable String maxSpread,
        @Nullable String fixed,
        @Nullable String min,
        @Nullable String max,
        @Nullable String mode
    ) {
        if (base == null) {
            warning("未指定基础值");
            return null;
        }
        if (spread == null) {
            warning("未指定浮动单位");
            return null;
        }
        if (maxSpread == null) {
            warning("未指定浮动范围上限");
            return null;
        }
        val baseValue = parse(base, 0D, NumberParser::parseDouble, "{0} 并非数字, 无法用作基础值");
        if (baseValue == null) return null;
        val spreadValue = parse(spread, 0D, NumberParser::parseDouble, "{0} 并非数字, 无法用作浮动单位");
        if (spreadValue == null) return null;
        val maxSpreadValue = parse(maxSpread, 0D, NumberParser::parseDouble, "{0} 并非数字, 无法用作浮动范围上限");
        if (maxSpreadValue == null) return null;

        double random = ThreadLocalRandom.current().nextGaussian() * spreadValue;
        random = Math.max(-maxSpreadValue, random);
        random = Math.min(maxSpreadValue, random);

        double result = baseValue * (1 + random);

        if (min != null) {
            val minValue = parse(min, 0D, NumberParser::parseDouble, "{0} 并非数字, 无法用作最小值");
            if (minValue == null) return null;
            result = Math.max(result, minValue);
        }

        if (max != null) {
            val maxValue = parse(max, 0D, NumberParser::parseDouble, "{0} 并非数字, 无法用作最大值");
            if (maxValue == null) return null;
            result = Math.min(result, maxValue);
        }

        val fixedValue = parse(fixed, 0, NumberParser::parseInteger, "{0} 并非数字, 无法用作取整位数");
        if (fixedValue == null) return null;

        val roundingMode = ScriptUtils.toRoundingMode(mode);

        return new BigDecimal(Double.toString(result)).setScale(fixedValue, roundingMode).toString();
    }
}

File: src/main/kotlin/pers/neige/neigeitems/section/impl/GaussianParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.SectionUtils.parseSection
import java.util.*
import kotlin.collections.HashMap

/**
 * gaussian节点解析器
 */
object GaussianParser : SectionParser() {
    private val RANDOM = Random()

    override val id: String = "gaussian"

    override fun onRequest(
        data: ConfigurationSection,
        cache: HashMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String? {
        return handler(cache, player, sections, true, data.getString("base"), data.getString("spread"), data.getString("maxSpread"), data.getString("fixed"), data.getString("min"), data.getString("max"))
    }

    override fun onRequest(
        args: List<String>,
        cache: HashMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        return handler(cache, player, sections, false, *args.toTypedArray()) ?: "<$id::${args.joinToString("_")}>"
    }

    private fun handler(cache: HashMap<String, String>?,
                        player: OfflinePlayer?,
                        sections: ConfigurationSection?,
                        parse: Boolean,
                        vararg args: String?): String? {
        // 获取基础数值
        val base = args.getOrNull(0)?.parseSection(parse, cache, player, sections)?.toDoubleOrNull()
        // 获取浮动单位
        val spread = args.getOrNull(1)?.parseSection(parse, cache, player, sections)?.toDoubleOrNull()
        // 获取浮动范围上限
        val maxSpread = args.getOrNull(2)?.parseSection(parse, cache, player, sections)?.toDoubleOrNull()
        // 获取取整位数
        val fixed = args.getOrNull(3)?.parseSection(parse, cache, player, sections)?.toIntOrNull() ?: 1
        // 获取大小范围
        val min = args.getOrNull(4)?.parseSection(parse, cache, player, sections)?.toDoubleOrNull()
        val max = args.getOrNull(5)?.parseSection(parse, cache, player, sections)?.toDoubleOrNull()
        // 获取随机数
        if (base != null && spread != null && maxSpread != null) {
            // 根据正态分布进行范围随机
            val random = (RANDOM.nextGaussian()*spread)
                // 限制随机范围下限
                .coerceAtLeast(-maxSpread)
                // 限制随机范围上限
                .coerceAtMost(maxSpread)
            var result = base*(1 + random)
            // 限制结果下限
            min?.let { result = result.coerceAtLeast(it) }
            // 限制结果上限
            max?.let { result = result.coerceAtMost(it) }
            // 返回结果(基础数值+基础数值*浮动范围)
            return "%.${fixed}f".format(result)
        }
        return null
    }
}